### PR TITLE
Cache frame representation info; doubles speed of coordinate generation.

### PR DESCRIPTION
**Pull request by dmopalmer**
Here's another caching fix.  This doubles the speed of SkyCoordinate generation.

```
%timeit SkyCoord(5,5,unit='deg')
# Before (current master branch)
1000 loops, best of 3: 1.41 ms per loop
# After
1000 loops, best of 3: 689 µs per loop
```

It also speeds up #3323 Slow iteration over a SkyCoord Object from 2.57 s (already an improvement over the original 19 s) to 1.24 s.

### CODE REVIEW

**Comment by pllim** [comment]
Your changes broke the tests.

**Comment by dmopalmer** [comment]
So it does.

Do you have any recommendations of where to invalidate the cache when a new subclass of a frame is created?  I thought that would be handled by the Metaclass.

```
       class FakeICRS(ICRS):
            frame_specific_representation_info = {
                'spherical': {'names': ('ra', 'dec', 'distance'),
                              'units': (None, None, None)},
                'unitspherical': {'names': ('ra', 'dec'),
                                  'units': (None, None)}
            }
    
            frame_specific_representation_info = {
                'spherical': [RepresentationMapping('lon', 'ra', u.hourangle),
                              RepresentationMapping('lat', 'dec', None),
                              RepresentationMapping('distance', 'distance')]  # should fall back to default of None unit
            }
            frame_specific_representation_info['unitspherical'] = \
                frame_specific_representation_info['spherical']
```

**Comment by pllim** [comment]
Maybe @eteq knows?

File: astropy/coordinates/baseframe.py [hunk 1]
```diff
@@ -789,6 +790,11 @@ def _get_representation_info(cls):
         # This exists as a class method only to support handling frame inputs
```
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: Seeing this, I wondered generally whether this still needed to be a class method. A quick test showed no failures at all when one just moves this to the `representation_info` property. Hence, it feels like this should be separated correctly in parts that are class and self-specific.

File: astropy/coordinates/baseframe.py [hunk 2]
```diff
@@ -789,6 +790,11 @@ def _get_representation_info(cls):
...
         # without units, which are deprecated and will be removed.  This can be
         # moved into the representation_info property at that time.
+        try:
+            if cls._representation_info_cachetick == MetaBaseRepresentation.REPRESENTATION_CLASSES_CACHE_TICK:
+                return cls._repr_attrs
+        except:
+            pass
 
         repr_attrs = {}
```
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: This first stanza would seem to have no business being on the class at all, as it only deals with representations. Might it make sense to move this to `representation.py`, perhaps adding a new dict with this information, and updating it on the metaclass?

**Comment by dmopalmer** [comment]
I don't really understand what it is used for and why.  (All I know is that a complex nested loop is called 22 times every time you want to represent the most basic element of positional astronomy.)

Moving it to somewhere else would require a deeper understanding of the code, its structure, and its history and customs than I have time to acquire at the moment.

I have no objection to someone else providing a better (and working) speed-up.

**Comment by mhvk** [comment]
@dmopalmer - fair enough; since I also do not understand all that well why this code is needed or called so often, maybe best is to ping someone else... @eteq: Could you have a look?

**Comment by mhvk** [comment]
Ideally, this is still tackled, but in pieces.

**Comment by eteq** [comment]
I'm reopening this but milestoning it for *3.1* - slated for that is a major re-write of the SkyCoord initializer in the name of performance.  It may be this becomes irrelevant at that time, in which case we should close it.

**Comment by eteq** [comment]
To answer the underlying question of why it's called so often: getting the representation info is needed to do all kinds of things like asking "what are the names of the attributes used on this frame". That's done in the `SkyCoord` initializer because it does a lot of "taking apart" an input object and then "putting it back together again".  One of the goals in the v3.1 optimization would be to undo that, possibly making this cache less/unnecessary.

But more generally, the problem with caching this in the manner this PR does is that in theory it can *change*.  I'm not sure if anyone actually uses this, but it *might* happen in some intermediate manner during class creation? My *suspicion* is that this is the origin of the failure.  But some slightly cleverer caching should still be possible...

**Comment by mhvk** [comment]
See #6941 for a PR that improves item access by a factor 50.  There is definitely low-hanging fruit that should get picked before we add more layers of caching (though this particular layer does still seem reasonable).

**Comment by eteq** [comment]
@dmopalmer - while this issue sat for a long while, in #7949 I finally cracked the case and got this caching working.  While it was not directly based on this PR (due to a lot of changes required for velocities that made it hard to do so), it was nonetheless inspired by your attempt here.  So I just now pushed up a "no-op" merge commit to this branch, which will enable us to merge this PR to record the commits in the git history so that we can be sure you get credit for the effort here that led to #7949 .

Thanks for the work here!

**Pull request merged**